### ISSUE TITLE
Retrieve Task Under Windows

### ISSUE THREAD

**Issue by thegogz** [ejholmes/metaforce#21]
Hi,

I'm having an issue with both the retrieve and deploy rake tasks when running under windows.

I've also tested this using the metaforce-demo project with the same results. 

Whenever I retrieve anything I get:

```
rake aborted!
can't dup NilClass
```

I'm not 100% sure but I think this may be related to RubyZip on Windows. Any ideas or pointers? 

Also how do I stop the soap being outputted to the console? I've tried setting Log: false in my yml file...

Thanks for your time,

Eoin

**Comment by ejholmes** [comment]
Can you provide the output using `rake --trace`?

You can disabled the soap messages with `Savon.log = false`.

**Comment by thegogz** [comment]
Here you go... with my password removed. Including Soap traces so everything looks like its working its just when it gets to the last step.

```
** Invoke retrieve_sandbox (first_time)
** Execute retrieve_sandbox
D, [2012-08-27T17:59:07.529069 #6836] DEBUG -- : Logging in via https://login.salesforce.com/services/Soap/u/23.0
D, [2012-08-27T17:59:07.535069 #6836] DEBUG -- : SOAP request: https://login.salesforce.com/services/Soap/u/23.0
D, [2012-08-27T17:59:07.535069 #6836] DEBUG -- : SOAPAction: "login", Content-Type: text/xml;charset=UTF-8, Content-Leng
th: 559
D, [2012-08-27T17:59:07.535069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><env:Envelope xmlns:xsd="http://w
ww.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="urn:partner.soap.sforce.com"
xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ins0="urn:partner.soap.sforce.com" xmlns:ins1="urn:sobject.p
artner.soap.sforce.com" xmlns:ins2="urn:fault.partner.soap.sforce.com"><env:Body><ins0:login><ins0:username>thegogz@gmai
l.com</ins0:username><ins0:password>*****************</ins0:password></ins0:login></env:Body></env:E
nvelope>
D, [2012-08-27T17:59:10.333069 #6836] DEBUG -- : SOAP response (status 200):
D, [2012-08-27T17:59:10.334069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="
http://schemas.xmlsoap.org/soap/envelope/" xmlns="urn:partner.soap.sforce.com" xmlns:xsi="http://www.w3.org/2001/XMLSche
ma-instance"><soapenv:Body><loginResponse><result><metadataServerUrl>https://na9-api.salesforce.com/services/Soap/m/23.0
/00DE0000000cU0q</metadataServerUrl><passwordExpired>false</passwordExpired><sandbox>false</sandbox><serverUrl>https://n
a9-api.salesforce.com/services/Soap/u/23.0/00DE0000000cU0q</serverUrl><sessionId>00DE0000000cU0q!AQoAQIpg0WjdcbbOmVI3ShC
i17wb1yrV7VJq_pKNadQRGelDG2wcxebB01lEezpm5tQ9bX8umlN4a9pookQfLpAKsicCuBrS</sessionId><userId>005E0000001ejPQIAY</userId>
<userInfo><accessibilityMode>false</accessibilityMode><currencySymbol>???</currencySymbol><orgAttachmentFileSizeLimit>52
42880</orgAttachmentFileSizeLimit><orgDefaultCurrencyIsoCode>EUR</orgDefaultCurrencyIsoCode><orgDisallowHtmlAttachments>
false</orgDisallowHtmlAttachments><orgHasPersonAccounts>false</orgHasPersonAccounts><organizationId>00DE0000000cU0qMAE</
organizationId><organizationMultiCurrency>false</organizationMultiCurrency><organizationName>Gmail</organizationName><pr
ofileId>00eE0000000vNwcIAE</profileId><roleId xsi:nil="true"/><sessionSecondsValid>7200</sessionSecondsValid><userDefaul
tCurrencyIsoCode xsi:nil="true"/><userEmail>thegogz@gmail.com</userEmail><userFullName>Eoin O&apos;Neill</userFullName><
userId>005E0000001ejPQIAY</userId><userLanguage>en_US</userLanguage><userLocale>en_IE_EURO</userLocale><userName>thegogz
@gmail.com</userName><userTimeZone>Europe/Dublin</userTimeZone><userType>Standard</userType><userUiSkin>Theme3</userUiSk
in></userInfo></result></loginResponse></soapenv:Body></soapenv:Envelope>
D, [2012-08-27T17:59:10.345069 #6836] DEBUG -- : Executing retrieve
D, [2012-08-27T17:59:10.404069 #6836] DEBUG -- : SOAP request: https://na9-api.salesforce.com/services/Soap/m/23.0/00DE0
000000cU0q
D, [2012-08-27T17:59:10.404069 #6836] DEBUG -- : SOAPAction: "retrieve", Content-Type: text/xml;charset=UTF-8, Content-L
ength: 841
D, [2012-08-27T17:59:10.405069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><env:Envelope xmlns:xsd="http://w
ww.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="http://soap.sforce.com/2006/0
4/metadata" xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ins0="http://soap.sforce.com/2006/04/metadata"><
env:Header><ins0:SessionHeader><ins0:sessionId>00DE0000000cU0q!AQoAQIpg0WjdcbbOmVI3ShCi17wb1yrV7VJq_pKNadQRGelDG2wcxebB0
1lEezpm5tQ9bX8umlN4a9pookQfLpAKsicCuBrS</ins0:sessionId></ins0:SessionHeader></env:Header><env:Body><ins0:retrieve><ins0
:retrieveRequest><ins0:apiVersion>23.0</ins0:apiVersion><ins0:singlePackage>true</ins0:singlePackage><ins0:unpackaged><i
ns0:types><tns:members>myclass</tns:members><tns:name>ApexClass</tns:name></ins0:types></ins0:unpackaged></ins0:retrieve
Request></ins0:retrieve></env:Body></env:Envelope>
D, [2012-08-27T17:59:11.457069 #6836] DEBUG -- : SOAP response (status 200):
D, [2012-08-27T17:59:11.457069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="
http://schemas.xmlsoap.org/soap/envelope/" xmlns="http://soap.sforce.com/2006/04/metadata"><soapenv:Body><retrieveRespon
se><result><done>false</done><id>04sE00000013LmSIAU</id><state>InProgress</state></result></retrieveResponse></soapenv:B
ody></soapenv:Envelope>
D, [2012-08-27T17:59:11.463069 #6836] DEBUG -- : Polling server for status on 04sE00000013LmSIAU
D, [2012-08-27T17:59:11.470069 #6836] DEBUG -- : SOAP request: https://na9-api.salesforce.com/services/Soap/m/23.0/00DE0
000000cU0q
D, [2012-08-27T17:59:11.471069 #6836] DEBUG -- : SOAPAction: "checkStatus", Content-Type: text/xml;charset=UTF-8, Conten
t-Length: 631
D, [2012-08-27T17:59:11.471069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><env:Envelope xmlns:xsd="http://w
ww.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="http://soap.sforce.com/2006/0
4/metadata" xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ins0="http://soap.sforce.com/2006/04/metadata"><
env:Header><ins0:SessionHeader><ins0:sessionId>00DE0000000cU0q!AQoAQIpg0WjdcbbOmVI3ShCi17wb1yrV7VJq_pKNadQRGelDG2wcxebB0
1lEezpm5tQ9bX8umlN4a9pookQfLpAKsicCuBrS</ins0:sessionId></ins0:SessionHeader></env:Header><env:Body><ins0:checkStatus><t
ns:ids>04sE00000013LmSIAU</tns:ids></ins0:checkStatus></env:Body></env:Envelope>
D, [2012-08-27T17:59:12.340069 #6836] DEBUG -- : SOAP response (status 200):
D, [2012-08-27T17:59:12.340069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="
http://schemas.xmlsoap.org/soap/envelope/" xmlns="http://soap.sforce.com/2006/04/metadata"><soapenv:Body><checkStatusRes
ponse><result><done>true</done><id>04sE00000013LmSIAU</id><state>Completed</state><stateDetailLastModifiedDate>2012-08-2
7T16:59:12.000Z</stateDetailLastModifiedDate></result></checkStatusResponse></soapenv:Body></soapenv:Envelope>
D, [2012-08-27T17:59:12.351069 #6836] DEBUG -- : Polling server for status on 04sE00000013LmSIAU
D, [2012-08-27T17:59:12.358069 #6836] DEBUG -- : SOAP request: https://na9-api.salesforce.com/services/Soap/m/23.0/00DE0
000000cU0q
D, [2012-08-27T17:59:12.359069 #6836] DEBUG -- : SOAPAction: "checkRetrieveStatus", Content-Type: text/xml;charset=UTF-8
, Content-Length: 647
D, [2012-08-27T17:59:12.359069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><env:Envelope xmlns:xsd="http://w
ww.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:tns="http://soap.sforce.com/2006/0
4/metadata" xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ins0="http://soap.sforce.com/2006/04/metadata"><
env:Header><ins0:SessionHeader><ins0:sessionId>00DE0000000cU0q!AQoAQIpg0WjdcbbOmVI3ShCi17wb1yrV7VJq_pKNadQRGelDG2wcxebB0
1lEezpm5tQ9bX8umlN4a9pookQfLpAKsicCuBrS</ins0:sessionId></ins0:SessionHeader></env:Header><env:Body><ins0:checkRetrieveS
tatus><tns:ids>04sE00000013LmSIAU</tns:ids></ins0:checkRetrieveStatus></env:Body></env:Envelope>
D, [2012-08-27T17:59:13.549069 #6836] DEBUG -- : SOAP response (status 200):
D, [2012-08-27T17:59:13.550069 #6836] DEBUG -- : <?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="
http://schemas.xmlsoap.org/soap/envelope/" xmlns="http://soap.sforce.com/2006/04/metadata"><soapenv:Body><checkRetrieveS
tatusResponse><result><fileProperties><createdById>005E0000001ejPQIAY</createdById><createdByName>Eoin O&apos;Neill</cre
atedByName><createdDate>2012-08-27T16:51:29.000Z</createdDate><fileName>classes/myclass.cls</fileName><fullName>myclass<
/fullName><id>01pE0000001MIkSIAW</id><lastModifiedById>005E0000001ejPQIAY</lastModifiedById><lastModifiedByName>Eoin O&a
pos;Neill</lastModifiedByName><lastModifiedDate>2012-08-27T16:51:34.000Z</lastModifiedDate><manageableState>unmanaged</m
anageableState><type>ApexClass</type></fileProperties><fileProperties><createdById>005E0000001ejPQIAY</createdById><crea
tedByName>Eoin O&apos;Neill</createdByName><createdDate>2012-08-27T16:59:12.204Z</createdDate><fileName>package.xml</fil
eName><fullName>package.xml</fullName><id></id><lastModifiedById>005E0000001ejPQIAY</lastModifiedById><lastModifiedByNam
e>Eoin O&apos;Neill</lastModifiedByName><lastModifiedDate>2012-08-27T16:59:12.204Z</lastModifiedDate><manageableState>un
managed</manageableState><type>Package</type></fileProperties><id>04sE00000013LmSIAU</id><zipFile>UEsDBBQACAAIAGaHG0EAAA
AAAAAAAAAAAAATAAAAY2xhc3Nlcy9teWNsYXNzLmNscysoTcrJTFZIzkksLlbIrQTT1VxctQBQSwcIJsTMJxYAAAAYAAAAUEsDBBQACAAIAGaHG0EAAAAAAA
AAAAAAAAAcAAAAY2xhc3Nlcy9teWNsYXNzLmNscy1tZXRhLnhtbE2NQQrCMBBF9zlFyN5MLCoiaUoRPIG6H9KogTYJnbH0+BYq4t+9z4Nnm3no5RRGijnVaq
uNkiH53MX0rNXtetkcVeOEbUuYzz0SycVPVKsXczkBUMai6ZFHH7TPA1TGHMDsYAiMHTIqJ+QyiyXe14ir9tpY+DtWgxj5Ta71HKdg4YvCwi/txAdQSwcIki
N/4IgAAACuAAAAUEsDBBQACAAIAGaHG0EAAAAAAAAAAAAAAAALAAAAcGFja2FnZS54bWxNjssKwjAQRff9ipC9mVhFRNIUEVy70A8Y01GLzYMmSPv3lj7QWc
0ZDneuKjvbsA+1sfau4GshOSNnfFW7Z8Fv1/Nqz0udqQuaNz6JDbaLBX+lFA4A0WMQ8eFbQ8J4C7mUO5BbsJSwwoRcZ2wYlfpAcdpHtmTvw0tte9NgjAqWw0
9xaEkfA3WnyRh5SoO/ODVX1/lGSAULZQrmxjr7AlBLBwj70rZLngAAAOMAAABQSwECFAAUAAgACABmhxtBJsTMJxYAAAAYAAAAEwAAAAAAAAAAAAAAAAAAAA
AAY2xhc3Nlcy9teWNsYXNzLmNsc1BLAQIUABQACAAIAGaHG0GSI3/giAAAAK4AAAAcAAAAAAAAAAAAAAAAAFcAAABjbGFzc2VzL215Y2xhc3MuY2xzLW1ldG
EueG1sUEsBAhQAFAAIAAgAZocbQfvStkueAAAA4wAAAAsAAAAAAAAAAAAAAAAAKQEAAHBhY2thZ2UueG1sUEsFBgAAAAADAAMAxAAAAAACAAAAAA==</zipF
ile></result></checkRetrieveStatusResponse></soapenv:Body></soapenv:Envelope>
rake aborted!
can't dup NilClass
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `dup'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `block in dup'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `map'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `dup'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:75:in `initialize'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:87:in `new'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:87:in `open'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/metaforce-0.5.3/lib/metaforce/metadata/transaction.rb:56:in `to'
C:/Ruby193/lib/ruby/gems/1.9.1/gems/metaforce-0.5.3/lib/metaforce/rake/retrieve.rb:31:in `block in define'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:205:in `call'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:200:in `each'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
C:/Ruby193/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
C:/Ruby193/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:116:in `invoke_task'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:94:in `block (2 levels) in top_level'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:94:in `each'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:94:in `block in top_level'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:88:in `top_level'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:66:in `block in run'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:63:in `run'
C:/Ruby193/bin/rake:32:in `<main>'
Tasks: TOP => retrieve_sandbox
```

**Comment by thegogz** [comment]
I've forked the repo and continued with some local testing. 

The issue is in the Transaction.to method 

Here's the stack when I replicate that method in a script locally. 

```
E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `dup': can't dup NilClass (TypeError)
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `block in dup'
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `map'
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_entry_set.rb:44:in `dup'
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:75:in `initialize'
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:87:in `new'
        from E:/Ruby192/lib/ruby/gems/1.9.1/gems/rubyzip-0.9.9/lib/zip/zip_file.rb:87:in `open'
        from ./metaforce_test.rb:19:in `<main>'
```

**Comment by ejholmes** [comment]
Hmm, yeah it's either an issue with ruby zip on windows or the call to Tempfile.new. I don't have a windows box I can test on unfortunately, but if you fix the issue and add tests, feel free to open a PR.

**Comment by thegogz** [comment]
I've found 2 issues but I've no idea how to write tests for either but to summarize for others (and if I find a way to write tests a PR) 

Tempfile.New opens the file in Text mode which leaves an extra newline character at the end of the file, not an issue for Unix systems but for windows this causes the binary file written to be corrupt. 

There's a simple workaround though basically setting the file to binary mode after is opened. 

``` ruby
f = Tempfile.new "binary"
f.binmode # switch to binary mode
```

The second issue I've found is more complex basically as suspected RubyZip has a compatibility issue with Windows I'm not sure why (i'd need to dig into RubyZip code to figure it out) the only short term solution I've found is to use ArchiveZip instead, I'm not sure how robust the ArchiveZip code is though, I've had a few issues where its silently erroring which isn't ideal.

However it does seem a bit cleaner, what I've done is changed the following code

``` ruby
Zip::ZipFile.open(path) do |zip|
  zip.each do |f|
    path = File.join(destination, f.name)
    FileUtils.mkdir_p(File.dirname(path))
    zip.extract(f, path) { true }
  end
end
```

to this 

``` ruby
Archive::Zip.extract(path, destination)
```
